### DISCUSSION TITLE
Dishwasher

### DISCUSSION THREAD

**Discussion by scaramatto** [ofalvai/home-assistant-candy#42] [Device support improvements]
Hi, 
I can't integrate my dishwasher in home assistant:
I have this error:

Config entry 'Candy' for candy integration not ready yet: Error communicating with API: Exception('Unable to detect machine type from API response', {'statusDWash': {'StatoWiFi': '1', 'StatoDWash': '0', 'CodiceErrore': 'E0', 'StartStop': '0', 'Program': 'P1', 'OpzProg': 'p', 'DelayStart': '0', 'RemTime': '0', 'TreinUno': '0', 'Eco': '0', 'MetaCarico': '0', 'ExtraDry': '0', 'MissSalt': '0', 'MissRinse': '0', 'OpenDoor': '0', 'Reset': '0', 'FWver': 'L1.12'}}); Retrying in background

if i connect at site: http://x.x.x.x/http-read.json?encrypted=0 i view this parameters:
{ "statusDWash":{ "StatoWiFi":"1", "StatoDWash":"0", "CodiceErrore":"E0", "StartStop":"0", "Program":"P1", "OpzProg":"p", "DelayStart":"0", "RemTime":"0", "TreinUno":"0", "Eco":"0", "MetaCarico":"0", "ExtraDry":"0", "MissSalt":"0", "MissRinse":"0", "OpenDoor":"0", "Reset":"0", "FWver":"L1.12" } }

Have any suggestions?

Thank you!!!

**Comment by ofalvai** [comment]
Hey @scaramatto 

I can add some basic support for dishwashers based on the API response you posted. But I would need a bit more information on how to parse the values and what they mean in practice, such as:

- `StatoDWash`: what are the possible numbers here and what is their meaning? For washing machines, I came up with the following states: https://github.com/ofalvai/home-assistant-candy/blob/main/custom_components/candy/client/model.py#L6
- `RemTime` is probably the remaining time, but is this seconds or minutes?
- `Eco`: I assume 0 is off and 1 is when eco mode is active?
- `ExtraDry`: same?

I'd appreciate any other observation about the meaning of the fields.

**Comment by scaramatto** [comment]
@ofalvai Thank for your answer! 

I think these values:

```
StartStop 0, 1

Program P0, P1, P2, P3, P4 

OpzProg 0, p, m,  (" ", +, -)

DelayStart value in minute

RemTime value in minute

TreinUno 0, 1 

Eco 0, 1

MetaCarico 0, 1

ExtraDry 0, 1

MissSalt 0, 1

MissRinse 0, 1

OpenDoor 0, 1

Reset 0, 1

Programs available: P0, P1, P1+, P1-, P2, P2+, P2-, P3, P3+, P3-, P4, P4+, P4-.

Example
Program P1+ its Program: P1  OpzProg: p
Program P1- its Program: P1  OpzProg: m
Program P1  its Program: P1  OpzProg: 0
```

I'm available for any test :)

**Comment by ofalvai** [comment]
Thank you, one last question before creating a PR for dishwasher support: what is the meaning of the values of `StatoDWash`? I see that it's `0` in the response you posted, but I've never seen a zero value on my washing machine. The idle state is marked with `1` in case of washing machines. It would be interesting to see the possible values and their meaning on dishwashers.

**Comment by scaramatto** [comment]
Hi, sorry for late response!

"StatoDWash":"0" **idle**
"StatoDWash":"1" **pre-wash**
"StatoDWash":"2" **wash**
"StatoDWash":"3" **rinse**
"StatoDWash":"4" **drying**    
"StatoDWash":"5" **Finished**

Thank for your support!

**Comment by ofalvai** [comment]
Thanks for the investigation @scaramatto, based on the API responses, I added initial support for dishwashers in #26

**Comment by IrealiTY** [comment]
hey hi,
some states might be not the same for all machines it seems.

I have HDPN 4S603PX DishWasher, State 3 here seems to be a Pre-Wash

![image](https://user-images.githubusercontent.com/1223686/140352928-7f675dd7-024e-4fce-94be-421127e96800.png)


```
{
        "statusDWash":
        {
                "StatoWiFi":"0",
                "StatoDWash":"3",
                "CodiceErrore":"E0",
                "StartStop":"1",
                "Program":"P8",
                "OpzProg":"0",
                "DelayStart":"0",
                "RemTime":"193",
                "TreinUno":"0",
                "Eco":"0",
                "MetaCarico":"0",
                "ExtraDry":"0",
                "OpenDoorOpt":"1",
                "MissSalt":"0",
                "MissRinse":"0",
                "OpenDoor":"0",
                "Reset":"0",
                "CheckUp":"0",
                "r1":"1",
                "r2":"1",
                "r3":"1",
                "r4":"1",
                "r5":"1",
                "r6":"0",
                "r7":"3",
                "r8":"0",
                "r9":"0",
                "D1":"0",
                "D1mac":"0"
        }
}
```

EDIT: Seems like this Dishwasher might not send the correct status? Its now on "Last Rinse" when I check the app but the StatoDWash is still on 3

**Comment by ofalvai** [comment]
@IrealiTY that's weird, and I hope that we can somehow come up with a universal response parsing because there is no API to get the model number and do model-based custom rules.

Let me know if you figured out what's happening with your device :) Maybe the app displays the status based on another property, not just `StatoDWash`?

**Comment by IrealiTY** [comment]
I'll try to log a few entries next time I use the machine
